File: Soluciones/MoscaVector/src/mosca/Mosca.java
package mosca;

import java.util.Scanner;

public class Mosca {
    public static void main(String[] args){
        System.out.println("La Mosca Vector");
        int numCasillas = 10;
        final int MOSCA = 1;


        // Pedir tamaño de vector
        numCasillas = pedirNumeroCasillas();

        // Crear el vector
        int[] casillas = new int[numCasillas];

        // Situar la mosca en el vector
        situarMosca(casillas, MOSCA);
        imprimirCasillas(casillas);

        boolean estaMuerta = false;
        int posMosca = 0;
        do {
            // Pedirle la posición donde vas a dar un tortazo
            posMosca = posicionGolpear(casillas);
            // Anlizamos el Golpeo
            estaMuerta = analizarGolpeo(MOSCA, casillas, estaMuerta, posMosca);
        } while(!estaMuerta);
        System.out.println("¡¡Has cazado a la maldita mosca!!!");
        System.out.println("Estaba en: " + (posMosca+1));

    }

    private static boolean analizarGolpeo(int MOSCA, int[] casillas, boolean estaMuerta, int posMosca) {
        // Logica del juego
        // Acertamos
        if(casillas[posMosca] == MOSCA) {
            estaMuerta = true;

            // Analizamos los limites
        } else {
            boolean checkIzda = posMosca != 0 && casillas[posMosca -1]== MOSCA;
            boolean checkDcha = posMosca != casillas.length-1 && casillas[posMosca +1]== MOSCA;

            if(checkIzda || checkDcha) {
                // Revoloteamos
                System.out.println("¡CASI!");
                situarMosca(casillas, MOSCA);
            }
        }

        if(!estaMuerta) {
            System.out.println("¡Has fallado!");
        }
        imprimirCasillas(casillas);
        return estaMuerta;
    }

    private static int posicionGolpear(int[] casillas) {
        Scanner sc = new Scanner(System.in);
        int posMosca;
        do {
            System.out.println("Introduce la posición a atacar: ");
            posMosca = sc.nextInt();
            System.out.println("La posición elegida es: " + posMosca);
        } while (posMosca<0 || posMosca>= casillas.length);
        return (posMosca-1);
    }

    private static void situarMosca(int[] casillas, int mosca) {
        int pos = 0;
        for (int i = 0; i < casillas.length; i++) {
            casillas[i] = 0;
        }
        do {
            pos = (int) (Math.random() * casillas.length);
            System.out.println("Posicion de la mosca: " + (pos+1));
        } while (pos<0 || pos>=casillas.length);
        casillas[pos] = mosca;
    }

    /**
     *
     * @return
     */
    private static int pedirNumeroCasillas() {
        Scanner sc = new Scanner(System.in);
        int numCasillas;
        do {
            System.out.println("Dime el número de casillas, siempre mayor a 5");
            numCasillas = sc.nextInt();
            System.out.println("Tamaño de vector es:" + numCasillas);
        } while (numCasillas<5);
        return numCasillas;
    }

    static void imprimirCasillas(int[] casillas) {
        System.out.print("{ ");
        for(int casilla: casillas){
            System.out.print(casilla + " ");
        }
        System.out.println("}");
    }
}

File: Soluciones/MoscaMatriz-Kotlin/src/main/kotlin/mosca/Mosca.kt
package mosca

import java.lang.Math.random


/**
 * Obtiene el número de Intentos
 */
fun getNumeroIntentos(): Int {
    var intentos = 0
    do {
        println("Introduce el número de intentos máximos, mayor a 1")
        // Si no lo puede comveretir da null
        intentos = readLine()?.toIntOrNull() ?: 0
    } while (intentos < 1)
    return intentos
}

fun getNumeroCasillas(): Int {
    var numCasillas: Int
    do {
        println("Dime el número de casillas, siempre mayor a 5")
        numCasillas =readLine()?.toIntOrNull() ?: 0
        println("Tamaño de vector es:$numCasillas")
    } while (numCasillas < 5)
    return numCasillas
}

fun imprimirCasillas(casillas: Array<IntArray>) {
    for (fila in casillas) {
        print("{ ")
        for (columna in fila) {
            print("$columna ")
        }
        println("} ")
    }
}

fun iniciarCasillas(casillas: Array<IntArray>) {
    // i recorre todos los indices
    for (i in casillas.indices) {
        // j recorre todos los elementos de cada indice
        for (j in casillas[i].indices) {
            // Se inicializa el valor de cada casilla a 0
            casillas[i][j] = 0
        }
        //for (j in 0 until casillas[i].size) {
        //    casillas[i][j] = 0
        //}
    }
}

fun getPosicion(casillas: Array<IntArray>): Int {
    var fila: Int
    do {
        fila = (random() * casillas.size).toInt()
        //System.out.println(s + (fila + 1));
    } while (fila < 0 || fila >= casillas.size)
    return fila
}

fun situarMosca(casillas: Array<IntArray>, mosca: Int) {
    val fila: Int = getPosicion(casillas)
    val columna: Int = getPosicion(casillas)
    casillas[fila][columna] = mosca
}

fun posicionGolpear(casillas: Array<IntArray>): IntArray {
    val posMosca = intArrayOf(0, 0)
    do {
        println("Introduce la posición de la Fila a atacar: ")
        posMosca[0] = (readLine()?.toIntOrNull() ?: 0 ) - 1
        println("La posición Fila elegida es: " + (posMosca[0] + 1))
    } while (posMosca[0] < 0 || posMosca[0] >= casillas.size)
    do {
        println("Introduce la posición de la Columna a atacar: ")
        posMosca[1] = (readLine()?.toIntOrNull() ?: 0 ) - 1
        println("La posición Columna elegida es: " + (posMosca[1] + 1))
    } while (posMosca[1] < 0 || posMosca[1] >= casillas.size)
    return posMosca
}

fun analizarGolpeo(MOSCA: Int, casillas: Array<IntArray>, estaMuerta: Boolean, posMosca: IntArray): Boolean {
    // Logica del juego
    // Acertamos
    var estaMuerta = estaMuerta
    if (casillas[posMosca[0]][posMosca[1]] == MOSCA) {
        estaMuerta = true

        // Analizamos los limites
    }
    for (i in -1..1) {
        for (j in -1..1) {
            if (posMosca[0] + i >= 0 && posMosca[0] + i < casillas.size && posMosca[1] + j >= 0 && posMosca[0] + j < casillas.size) {
                if (casillas[posMosca[0] + i][posMosca[1] + j] == MOSCA) {
                    println("¡CASI!")
                    iniciarCasillas(casillas)
                    situarMosca(casillas, MOSCA)
                }
            }
        }
    }
    if (!estaMuerta) {
        println("¡Has fallado!")
    }
    imprimirCasillas(casillas)
    return estaMuerta
}

// Vamos a buscar la mosca
fun buscarMosca(mosca: Int = 1, casillas: Array<IntArray>): IntArray {
    for (i in casillas.indices) {
        for (j in casillas[i].indices) {
            if (casillas[i][j] == mosca) {
                //println("La mosca está en la posición: " + (i + 1) + "," + (j + 1))
                return intArrayOf(i, j)
            }
        }
    }
    return intArrayOf(-1, -1)
}

